tools/infrastructure/api_compare.py

```diff
@@ -0,0 +1,368 @@
+    'success': lambda rpc: rpc.rpc_type != "response"
+}
+
+exclude_params_from_mob_rpc = {
+    'type': (lambda rpc: rpc.rpc_name not in ["AddCommand", "DeleteCommand"]),
+    'appID': (lambda rpc: False),
+    'grammarID': (lambda rpc: False)}
+
+
+# Functions defining specific comparison rules for specificattributes
+
+def CheckMandatoryParam(attributes):
+    result = {}
+    minsize_exists = "minsize" in attributes
+    maxsize_exists = "maxsize" in attributes
+    if not minsize_exists:
+        attributes["minsize"] = None
+        result["array"] = attributes["array"]
+        result["minsize"] = attributes["minsize"]
+    if not maxsize_exists:
+        attributes["maxsize"] = None
+        result["array"] = attributes["array"]
+        result["maxsize"] = attributes["maxsize"]
+    return result
+
+
+# Comparison rule when attribute "type" = "String"
+def StringCompareRule(mob_param_attributes, hmi_param_attributes):
+    attr = "minlength"
+    if attr not in mob_param_attributes:
+        if attr not in hmi_param_attributes:
+            return {attr: None}, {attr: None}
+        elif hmi_param_attributes[attr] != "1":
+            return {attr: None}, {attr: hmi_param_attributes[attr]}
+        else:
+            mob_param_attributes[attr] = "1"
+    return {}, {}
+
+
+# Comparison rule for attribute "array"
+def ArrayCompareRules(mob_param_attributes, hmi_param_attributes):
+    attr = "array"
+    mobile_result = {}
+    hmi_result = {}
+    if attr in mob_param_attributes and mob_param_attributes[attr] == "true":
+        mobile_result = CheckMandatoryParam(mob_param_attributes)
+    if attr in hmi_param_attributes and hmi_param_attributes[attr] == "true":
+        hmi_result = CheckMandatoryParam(hmi_param_attributes)
+    return mobile_result, hmi_result
+
+
+# Common comparison function
+def AllCompareRules(mob_param_attributes, hmi_param_attributes):
+    mobile_result = {}
+    hmi_result = {}
+    attr_names = mob_param_attributes.keys() + hmi_param_attributes.keys()
+    attr_names = set(attr_names)
+    for attr_name in attr_names:
+        mobile_attribute_value = None
+        hmi_attribute_value = None
+        compare_func = simple_values_comparison
+        if attr_name in attributes_comparison_rules:
+            # Get appropriate comparison function from attributes comparison
+            # rules
+            compare_func = attributes_comparison_rules[attr_name]
+        if attr_name in mob_param_attributes:
+            mobile_attribute_value = mob_param_attributes[attr_name]
+        if attr_name in hmi_param_attributes:
+            hmi_attribute_value = hmi_param_attributes[attr_name]
+        values_equal = compare_func(
+            mobile_attribute_value, hmi_attribute_value)
+        if not values_equal:
+            hmi_result[attr_name] = hmi_attribute_value
+            mobile_result[attr_name] = mobile_attribute_value
+    return mobile_result, hmi_result
+
+
+global_compare_rules = [
```

Comment:
@OHerasym I am bot sure that it's worth to add comments describing every comparison rule as there are many rules applied.